<div style="background-color:#e0f0ff; padding:15px; border-left:6px solid #3399ff; border-radius:4px;">

🔵 **Nota para el equipo de Tripleten**

Este proyecto ya había sido completado, pero **no aparece como calificado en mi perfil**.  
Posiblemente se perdió el registro al momento de cambiarme de cohorte, ya que ahora me aparece un **95% de avance**, aunque ya terminé todos los proyectos.

Por eso, **lo envío nuevamente para su revisión** y para que me ayuden a **actualizar el avance correcto en la plataforma**.

¡Muchas gracias por el apoyo!  
</div>

# Déjame escuchar música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa de ellos y tomar decisiones en consecuencia. Esto implica pasar por diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investiguemos, necesitamos formular hipótesis que después podamos probar. A veces, aceptaremos estas hipótesis; otras veces, las rechazaremos. Para tomar las decisiones adecuadas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de música online para probar la hipótesis que planteamos a continuación y compararás el comportamiento de los usuarios de estas dos ciudades.

### Objetivo:
Probar la hipótesis:
1. La actividad de los usuarios difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento de los usuarios se almacenan en el archivo `/datasets/music_project_en.csv`. No hay información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, debes evaluar la calidad de los datos y ver si los problemas son significativos. Más tarde, durante el preprocesamiento de datos, deberás abordar los problemas más críticos.

Tu proyecto contará con estas tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de los datos.
 3. Prueba de la hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importa pandas
import pandas as pd



Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Lee el archivo y almacénalo en df
# Cargar los datos
df = pd.read_csv('/datasets/music_project_en.csv')


Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtén las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
# Obtén la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?
Las filas representan registros individuales de reproducción de música por parte de usuarios en una plataforma.
Cada fila contiene información como el ID del usuario, la canción reproducida, el artista, el género musical, la ciudad desde la que se escuchó, la hora del día y el día de la semana.
Podemos saber qué almacena cada columna gracias a sus nombres y valores, además de utilizar funciones como .head(), .info() y .describe() para obtener una vista previa, tipos de datos y estadísticas generales.
2. ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?
Sí, hay suficientes datos para probar la hipótesis principal del proyecto.
El dataset contiene más de 65 mil registros, con usuarios repartidos entre dos ciudades y en tres días distintos. Esta cantidad permite realizar un análisis comparativo con pruebas estadísticas como la t de Student para observar si hay diferencias en el comportamiento de los usuarios por ciudad y día.

3.¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?
Sí, se detectaron los siguientes problemas durante el preprocesamiento:

Los encabezados de la tabla tienen diferente orden(Unos empiezan con mayuscolas)
El time esta como object

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
print(df.columns)


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


In [6]:
# Nombres de columnas como lista
df.columns.tolist()

['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day']

Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case.


Anteriormente, aprendiste una forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora.

Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [7]:
# Bucle en los encabezados que lo pone todo en minúsculas
df.columns = [col.lower() for col in df.columns]

Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [8]:
# Bucle en los encabezados que elimina los espacios
df.columns = [col.strip() for col in df.columns]

Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [9]:
# Cambia el nombre de la columna "userid"
df.rename(columns={'userid': 'user_id'}, inplace=True)

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [10]:
# Comprueba el resultado: lista de encabezados
df.columns.tolist()

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [11]:
# Calcula el número de valores ausentes
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

No todos los valores ausentes afectan de la misma forma a la investigación. Por ejemplo, los valores ausentes en 'track' y 'artist' no son cruciales para el análisis, ya que estos datos son más descriptivos que analíticos. Por eso, puedes reemplazarlos directamente con un valor predeterminado como el string 'unknown' (desconocido).

En cambio, los valores ausentes en 'genre' sí pueden influir en la comparación entre las preferencias musicales de Springfield y Shelbyville. En un escenario real, lo ideal sería investigar por qué faltan estos datos e intentar recuperarlos. Pero en este proyecto no tenemos esa posibilidad, así que deberás rellenar esos valores con un valor predeterminado.


Como vimos anteriormente en las lecciones, la mejor manera de hacerlo es crear una lista con los nombres de las columnas que necesitan reemplazarse.  Luego, utilizar esta lista para iterar sobre cada columna y realizar el reemplazo correspondiente.

Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [12]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
# Lista de columnas con valores ausentes que se quieren reemplazar
columns_to_replace = ['track', 'artist', 'genre']

# Reemplazar los valores ausentes en cada columna con 'unknown'
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [13]:
# Cuenta los valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [14]:
# Cuenta los duplicados explícitos
duplicados = df.duplicated().sum()
print(f"Total de filas duplicadas: {duplicados}")

Total de filas duplicadas: 3826


Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [15]:
# Elimina los duplicados explícitos
df = df.drop_duplicates()

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [16]:
# Comprueba de nuevo si hay duplicados
duplicados = df.duplicated().sum()
print(f"Total de filas duplicadas: {duplicados}")

Total de filas duplicadas: 0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [17]:
# Inspecciona los nombres de géneros únicos

# Extraer la columna 'genre' y obtener valores únicos ordenados
unique_genres = sorted(df['genre'].unique())

# Mostrar la lista
print(unique_genres)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Algunos de los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_genres()` que tendrá dos parámetros:

* `wrong_genres`: una lista con todos los valores que deben ser reemplazados.  
* `correct_genre`: un string que se utilizará como valor de reemplazo.

El objetivo de esta función es **corregir los valores en la columna `'genre'` del DataFrame `df`**, reemplazando cada valor de la lista `wrong_genres` por `correct_genre`.

Etapa 2.12.
1. Define una función llamada `replace_wrong_genres()` que reciba dos parámetros: `wrong_genres` y `correct_genre`.

2. Dentro de la función, utiliza un bucle `for` para iterar sobre cada valor en la lista `wrong_genres`.

3. En cada iteración, accede a la columna `'genre'` del DataFrame `df` y utiliza el método `.replace()` para sustituir el valor incorrecto por `correct_genre`.

4. Llama a la función y pasa como argumentos:
   - Una lista con los duplicados implícitos: `['hip', 'hop', 'hip-hop']`
   - El string de reemplazo: `'hiphop'`

In [18]:
# Función para reemplazar los duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for genre in wrong_genres:
        df['genre'] = df['genre'].replace(genre, correct_genre)



Etapa 2.13. Ahora, llama a `replace_wrong_genres()` y pásale estos argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los sustituya por `hiphop`:

In [19]:
# Elimina los duplicados implícitos

replace_wrong_genres(['hip', 'hop', 'hip-hop'], 'hiphop')

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [20]:
# Comprueba de nuevo los duplicados implícitos
unique_genres = sorted(df['genre'].unique())

# Mostrar la lista
print(unique_genres)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

[Volver a Contenidos](#back)

### Observaciones <a id='data_preprocessing_conclusions'></a>

Resumen del Análisis de Datos
Durante el análisis del conjunto de datos sobre preferencias musicales, se identificaron y corrigieron varios problemas que podrían haber afectado la calidad del análisis:

Problemas detectados:
Valores duplicados: Se encontraron 3,826 registros duplicados.
Valores ausentes: Las columnas track, artist y genre contenían datos nulos.
Inconsistencias en los nombres de columnas: Algunos encabezados tenían espacios innecesarios o mayúsculas inconsistentes.
Duplicados implícitos en el género musical: Existían variantes como 'hip', 'hop', y 'hip-hop' que se referían al mismo género.

Acciones realizadas:
Se eliminaron los duplicados para evitar sesgos en el conteo de actividad.
Se rellenaron los valores nulos con el string 'unknown' para mantener la estructura del DataFrame sin eliminar datos.
Se limpiaron los nombres de columnas, eliminando espacios y convirtiéndolos a minúsculas, lo que mejora la legibilidad y evita errores de codificación.
Se creó una función personalizada para unificar los géneros musicales duplicados implícitos, reemplazándolos por un único valor ('hiphop').

Mejora en la calidad del análisis:
Estas acciones garantizaron que los datos fueran más limpios, consistentes y confiables, permitiendo realizar análisis estadísticos precisos y comparaciones válidas entre ciudades y días de la semana. Al eliminar el ruido y las inconsistencias, el análisis final fue más robusto y fácil de interpretar.



Resumen del Análisis
Durante el análisis se detectaron duplicados, valores nulos y géneros musicales escritos de forma inconsistente.
Se eliminaron los registros duplicados, se reemplazaron los valores ausentes con 'unknown', y se unificaron géneros como 'hip', 'hop' y 'hip-hop' en 'hiphop'.

Estas correcciones mejoraron la calidad de los datos, haciéndolos más claros y consistentes para un análisis confiable.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de la hipótesis <a id='hypothesis'></a>

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

La hipótesis que queremos probar plantea que existen diferencias en la forma en que los usuarios de Springfield y Shelbyville consumen música.
Para analizar esto, nos enfocaremos en los datos correspondientes a tres días específicos de la semana: lunes, miércoles y viernes.

El análisis consistirá en comparar la cantidad de canciones reproducidas por los usuarios de cada ciudad en esos días. Esto nos permitirá observar posibles patrones o diferencias en los hábitos de consumo entre Springfield y Shelbyville.

Para llevar a cabo este análisis, es importante seguir el enfoque de dividir-aplicar-combinar, del que ya hablamos en la lección. En este caso:

*  Dividir: separamos los datos en grupos según la ciudad.

*  Aplicar: dentro de cada grupo, contamos cuántas canciones se reprodujeron.

*  Combinar: reunimos los resultados en una estructura que nos permita comparar fácilmente ambas ciudades.

Este procedimiento debe repetirse de forma independiente para cada uno de los tres días seleccionados. El resultado final debe mostrar el número de reproducciones por ciudad en cada uno de esos días.

Una posible forma de estructurar el código sería con la siguiente expresión:

`df.groupby(by='...')['...'].method()`

Deberás completar los argumentos correspondientes para agrupar por ciudad y contar las canciones reproducidas. Este enfoque te dará una visión clara y comparativa del comportamiento de los usuarios en ambas ciudades.

Etapa 3.1. Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna 'track' como referencia.

In [21]:
# Cuenta las canciones reproducidas en cada ciudad
# Contar cuántas canciones se reprodujeron en cada ciudad
reproducciones_por_ciudad = df.groupby('city')['track'].count()

# Mostrar el resultado
print(reproducciones_por_ciudad)

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


`Redacta brevemente tus observaciones sobre los resultados. ¿Qué diferencias notaste entre Springfield y Shelbyville? ¿A qué podrían deberse esas diferencias?`

Etapa 3.2.
Observaciones
Se registraron 42,741 reproducciones en Springfield y 18,512 en Shelbyville.
Esto indica que los usuarios de Springfield escucharon más del doble de canciones que los de Shelbyville durante los días analizados.

Estas diferencias podrían deberse a factores como:

Una mayor cantidad de usuarios activos en Springfield.

Diferencias culturales o de acceso a internet/dispositivos.

O simplemente hábitos de consumo distintos entre las dos ciudades.

Etapa 3.3.
1. Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes, miércoles y viernes.

2. Utiliza el mismo método de conteo que antes, pero ahora cambia la columna de agrupación para enfocarte en el día.

3. Esto te permitirá identificar posibles patrones de escucha según el día de la semana.


In [22]:
# Calcula las canciones reproducidas en cada uno de los tres días

# Contar cuántas canciones se reprodujeron por día de la semana
reproducciones_por_dia = df.groupby('day')['track'].count()

# Mostrar los resultados
print(reproducciones_por_dia)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


`Describe brevemente qué observaste al comparar los lunes, miércoles y viernes. ¿Hubo algún día con menos actividad? ¿Cambian las conclusiones si analizas cada ciudad por separado?`

Etapa 3.4. Observaciones por día
Al comparar los días de la semana:
Viernes fue el día con mayor número de reproducciones (21,840).
Miércoles tuvo la menor actividad (18,059).
El lunes se mantuvo intermedio (21,354).

Esto sugiere que los usuarios escuchan más música hacia el final de la semana.
Sin embargo, para sacar conclusiones más precisas, sería importante analizar los días por separado para cada ciudad, ya que es posible que los patrones de consumo varíen entre Springfield y Shelbyville.

Hasta ahora has aprendido a contar entradas agrupándolas por un solo criterio, como la ciudad o el día de la semana. Ahora vamos a dar un paso más: necesitas crear una función que **cuente cuántas canciones se reprodujeron en una ciudad específica durante un día determinado**, combinando ambos criterios de filtrado.

La función se llamará `number_tracks()` y debe aceptar dos parámetros:

- `day`: un día de la semana (por ejemplo, `'Monday'`).
- `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`).

Dentro de la función, deberás aplicar un **filtrado secuencial mediante indexación lógica**: primero tendrás que filtrar el DataFrame por el día, y luego por la ciudad. Una vez que tengas el subconjunto de datos correcto, debes contar cuántas veces aparece un valor en la columna `'user_id'`. Ese número representará el total de canciones reproducidas bajo esos dos criterios.



Etapa 3.5.
1. Declara una función llamada `number_tracks()` con dos parámetros: `day` y `city`.

2. Filtra el DataFrame para conservar solo las filas donde la columna `'day'` sea igual al valor del parámetro `day`.

3. A partir del resultado anterior, filtra nuevamente para conservar solo las filas donde la columna `'city'` sea igual al valor del parámetro `city`.

4. Extrae la columna `'user_id'` del DataFrame filtrado y utiliza el método `.count()` para contar el número de entradas.

5. Guarda ese valor en una variable y **devuélvelo** como resultado de la función.


In [23]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    day_filtered = df[df['day'] == day]
    
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    city_filtered = day_filtered[day_filtered['city'] == city]
    
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    count = city_filtered['user_id'].count()
    
    # Devuelve el número de valores de la columna 'user_id'
    return count


Etapa 3.6. Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que puedas recuperar los datos de ambas ciudades para cada uno de los tres días.

In [24]:
# El número de canciones reproducidas en Springfield el lunes
print(number_tracks(day='Monday', city='Springfield'))


15740


In [25]:
# El número de canciones reproducidas en Shelbyville el lunes

print(number_tracks(day='Monday', city='Shelbyville'))


5614


In [26]:

# El número de canciones reproducidas en Springfield el miércoles
print(number_tracks(day='Wednesday', city='Springfield'))



11056


In [27]:
# El número de canciones reproducidas en Shelbyville el miércoles
print(number_tracks(day='Wednesday', city='Shelbyville'))



7003


In [28]:
# El número de canciones reproducidas en Springfield el viernes
print(number_tracks(day='Friday', city='Springfield'))



15945


In [29]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks(day='Friday', city='Shelbyville'))

5895


[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

##Escribe tus conclusiones finales sobre la hipótesis.
 * ¿Los datos apoyan la idea de que el comportamiento de los usuarios respecto a la música que escuchan varía según la ciudad y el día de la semana?
 * Indica si la hipótesis debe aceptarse o rechazarse, y justifica tu respuesta con base en los resultados obtenidos.

Etapa 4. Conclusión Final
Los datos muestran diferencias claras en la cantidad de canciones reproducidas entre Springfield y Shelbyville en los tres días analizados (lunes, miércoles y viernes):
En todos los días, Springfield tuvo una actividad significativamente mayor que Shelbyville.
Por ejemplo, el lunes Springfield registró 15,740 reproducciones, mientras que Shelbyville solo 5,614.
Esto indica que el comportamiento de los usuarios sí varía según la ciudad y el día de la semana, apoyando parcialmente la hipótesis.

Veredicto:
La hipótesis puede aceptarse.
Los datos respaldan la idea de que hay diferencias en la actividad musical según la ciudad y el día, aunque sería necesario un análisis estadístico más profundo para determinar si estas diferencias son estadísticamente significativas.

### Nota
En los proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)